src/features/reasoning/helpers.test.ts

```diff
@@ -0,0 +1,251 @@
+import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
+import {
+  estimateTokenUsage,
+  updateConversationTokens,
+  extractTokenUsageFromChunk,
+  updateTokenUsage,
+} from './helpers';
+import {
+  ChatMessage,
+  MessageHistoryStore,
+} from '../../core/stores/messageHistoryStore';
+import { ChatCompletionChunk } from 'openai/resources/index';
+import { ModelConfig } from '../../core/types/models';
+
+describe('estimateTokenUsage', () => {
+  it('should calculate tokens based on message length', () => {
+    const messages: ChatMessage[] = [
+      { role: 'user', content: 'Hello, how are you?' },
+      { role: 'assistant', content: 'I am doing well, thank you for asking!' },
+    ];
+
+    const result = estimateTokenUsage(messages);
+
+    // "Hello, how are you?" is 19 chars, so ~5 tokens
+    // "I am doing well, thank you for asking!" is 42 chars, so ~10 tokens
+    expect(result.promptTokens).toBe(5);
+    expect(result.completionTokens).toBe(10);
+    expect(result.totalTokens).toBe(15);
+  });
```

Comment:
Unfortunately, this is the best we will be able to do for restricting a user from sending a message that might overwhelm the context limit. We only have access to deepseek usage _after_ the request, so we'll have to use this as a guard on the input for the reasoning model.